File: packages/flags/src/flags.ts
// =====================================================================
// 🚩 @qoe/flags — Registre typé des feature flags
// =====================================================================
// 📖 Source unique des feature flags du monorepo. Chaque clé déclare sa
//    VALEUR PAR DÉFAUT : c'est le fallback utilisé quand le flag est
//    éteint, non configuré dans GrowthBook, ou quand GrowthBook est
//    injoignable (dégradation gracieuse — jamais de crash).
//
// 🎯 Pour ajouter un flag :
//    1. Ajoute une entrée ici (clé kebab-case + défaut)
//    2. Crée le flag dans l'UI GrowthBook (Features → New Feature)
//       avec la même clé et la même valeur par défaut
//    3. Utilise-le côté client :  useFlag('ma-feature')
//       ou côté serveur :         isFlagOn('ma-feature', attrs)
// =====================================================================

export const FLAGS = {
  // 📰 Feed — carousel "Recommandations" de la home (démo feature flag)
  'feed-recommendations': true,
  // 🌐 Web — bandeau newsletter en bas du site public
  'web-newsletter-banner': false,
  // 🎨 Dashboard — suggestions de titres par IA dans l'éditeur
  'dashboard-ai-title-suggestions': false,
  // 🏝️ Landing — nouvelle section pricing
  'landing-pricing-section': false,
  // 🛡️ Admin — journal d'audit des actions admin
  'admin-audit-log': false,
  // ⚙️ Workers — coupe-feu sur l'envoi des newsletters (kill switch)
  'workers-newsletter-dispatch': true,
} as const satisfies Record<string, boolean>;

export type FlagKey = keyof typeof FLAGS;

/**
 * Retourne la valeur par défaut d'un flag (le fallback si éteint/indispo).
 */
export function defaultFor<K extends FlagKey>(key: K): (typeof FLAGS)[K] {
  return FLAGS[key];
}

File: apps/api/internal/flags/flags.go
// Package flags — feature flags lus côté serveur Go.
//
// Unification UI + sécurité : la table `feature_flags` (même Postgres que
// l'app, pilotée depuis la console admin via Supabase / @qoe/flags) est lue
// directement par l'API Go. Un flag activé/désactivé dans la console est donc
// enforceable côté serveur (workers, admin…) et exposable via GET /v1/flags —
// sans nouvelle écriture, une seule source de vérité.
//
// Dégradation gracieuse : table absente, DB down ou clé inconnue → valeur par
// défaut du registre (miroir exact de packages/flags/src/flags.ts). Zéro crash.
package flags

import (
	"context"
	"sync"
	"time"

	"github.com/jackc/pgx/v5/pgxpool"
)

// Clés des feature flags (kebab-case, miroir de @qoe/flags).
const (
	FeedRecommendations   = "feed-recommendations"
	WebNewsletterBanner   = "web-newsletter-banner"
	DashboardAITitle      = "dashboard-ai-title-suggestions"
	LandingPricingSection = "landing-pricing-section"
	AdminAuditLog         = "admin-audit-log"
	WorkersNewsletter     = "workers-newsletter-dispatch"
)

// defaults est le registre des défauts (miroir exact de @qoe/flags) : utilisé
// quand la table est absente, la clé inconnue, ou la DB injoignable.
var defaults = map[string]bool{
	FeedRecommendations:   true,
	WebNewsletterBanner:   false,
	DashboardAITitle:      false,
	LandingPricingSection: false,
	AdminAuditLog:         false,
	WorkersNewsletter:     true,
}

// cacheTTL borne la fraîcheur d'un flag : une bascule console est appliquée
// côté serveur en quelques secondes, sans redéploiement. Var (non const) pour
// être raccourci dans les tests.
var cacheTTL = 30 * time.Second

// Service lit les feature flags depuis la table partagée, avec cache TTL.
type Service struct {
	pool *pgxpool.Pool

	mu     sync.Mutex
	at     time.Time
	cached map[string]bool
}

func NewService(pool *pgxpool.Pool) *Service {
	return &Service{pool: pool}
}

// All retourne l'état de tous les flags (registre + surcharges table).
func (s *Service) All(ctx context.Context) map[string]bool {
	s.mu.Lock()
	defer s.mu.Unlock()

	out := map[string]bool{}
	for k, v := range defaults {
		out[k] = v
	}
	if time.Since(s.at) >= cacheTTL {
		s.reload(ctx)
	}
	for k, v := range s.cached {
		out[k] = v
	}
	return out
}

// IsOn évalue un flag (défaut du registre si clé inconnue / DB down).
func (s *Service) IsOn(ctx context.Context, key string) bool {
	if _, known := defaults[key]; !known {
		return false
	}
	return s.All(ctx)[key]
}

// reload rafraîchit le cache depuis la table feature_flags (best-effort).
// Doit être appelé avec s.mu tenue.
func (s *Service) reload(ctx context.Context) {
	s.cached = map[string]bool{}
	rows, err := s.pool.Query(ctx, `SELECT key, is_enabled FROM feature_flags`)
	if err != nil {
		// DB down → on garde le cache précédent (ou registre vide → défauts).
		if s.cached == nil {
			s.cached = map[string]bool{}
		}
		return
	}
	defer rows.Close()
	for rows.Next() {
		var key string
		var enabled bool
		if err := rows.Scan(&key, &enabled); err != nil {
			continue
		}
		if _, known := defaults[key]; known {
			s.cached[key] = enabled
		}
	}
	s.at = time.Now()
}
